### ISSUE TITLE
documentation: parallelization of the suite

### ISSUE THREAD

**Issue by Ulfgard** [numbbo/coco#707]
Hi,

I am wondering whether it is possible to parallelize the functions of the suite (e.g. going through each problem and starting a thread for each). This is because currently evaluation can take a long time and 94% of my cores are idle.

I am looking at this simple loop:

```
coco_suite_t* suite = coco_suite("bbob-biobj", "year: 2016", "dimensions: 5");
coco_observer_t* observer = coco_observer("bbob-biobj", "...");

coco_problem_t* problem;
while ((problem = coco_suite_get_next_problem(suite, observer)) != NULL) {
    //this while loop should be run in parallel
}
```

the problem is, that this is not a "nice" loop for easy parallelization with OpenMP. for that a for loop over all problem numbers would be much easier.

I am currently thinking about the following solution in C++:

```
//first query all problems
std::vector<coco_problem_t*> problems;
coco_problem_t* problem;
while ((problem = coco_suite_get_next_problem(suite, observer)) != NULL) {
    problems.push_back(problem);
}
//run them in parallel
PARALLEL_FOR(std::size_t i = 0; problems.size(); ++i){
    //...
}
```

But the documentation does not state whether the observer is thread safe, whether several independent observers are possible or whether the problem instances actually exist after the next call of of coco_suite_get_next_problem.

**Comment by ttusar** [comment]
Your are right, the documentation doesn't address this, we need to expand it. 

Your example won't work because the problem instance returned by `coco_suite_get_next_problem` is available only up to the next call to the same function. Your while loop can be substituted by this (note it's untested):
```C
for (std::size_t index = 0; index < coco_suite_get_number_of_problems(suite); index++) {
    problem = coco_suite_get_problem(suite, index);
    problems.push_back(problem);
}
```
However, the big question is what to do with the observers. The problems above are not observed and you need as many observers as you have parallel threads (this should work fine). I'm not sure what `PARALLEL_FOR` does exactly (I'm not familiar with OpenMP), but you need to wrap each problem with an observer using `coco_problem_add_observer()` before you run your algorithm on it. 

In the end you will also need to free all problems and all observers. It is enough if you free each problem using `coco_problem_free(problem)`, i.e., you don't need to explicitly remove the observer first using `coco_problem_remove_observer()`.

**Comment by nikohansen** [comment]
If we create a single observer instance for each problem, this is supposed to work, at least in my mind. For this very reason we tried to get rid of global/static variables in the observer implementation.

**Comment by ttusar** [comment]
@nikohansen The fact that `coco_suite_get_problem` can return NULL (when the problem has been filtered out from the suite) slipped under my radar, but does not create problems in the C code, because it wasn't used by the example experiment or any other COCO C code so far. Is the python interface affected by this (I will check the other languages)?

**Comment by ttusar** [comment]
I think Matlab is fine, while Java didn't interface this function up to now.

**Comment by Ulfgard** [comment]
i have played around with several observers yesterday, but I got weird error messages. 
I think that Niko is right, that every problem needs a different observer- or at least one observer per function. I am updating this once i got a fully working version

**Comment by nikohansen** [comment]
@ttusar according to my tests the selection seems to work fine. Reading the code I don't understand why, given the C code returns `NULL` once in a while.

**Comment by nikohansen** [comment]
@ttusar how can we differentiate between an "intermediate" `NULL` and the final `NULL`, after which no problem is left?

**Comment by Ulfgard** [comment]
Okay, just as an update. Given the code from @ttusar i was able to get the correct list of problems. However, i really ran into troubles using the observers, until i got the following code:

```
#pragma omp parallel for
for(std::size_t index = 0; index < problems.size(); index++) {
	coco_problem_t*& problem = problems[index];
	coco_observer_t* observer = observers[SHARK_THREAD_NUM];
	problem = coco_problem_add_observer(problem,observer);
	//now do some oiptimization with the problem
	coco_problem_free(problem);//problem will crash with error (1) without this line, even single threaded
}
```

Error (1) is:
COCO FATAL ERROR: logger_biobj_indicator() failed to open file 'exdata/F2-ArchiveMOCMA/Thread/2-moderate_5-weakly-structured_hyp.info'

it can also be a simple crash of the program.

replacing the line (1) with 
`coco_problem_remove_observer(problem,observer);`

COCO FATAL ERROR: 3 coco_problem_remove_observer(): trying to remove observer bbob-biobj instead of 1-separabl

**Comment by Ulfgard** [comment]
I am also struggling with the post-processing:

> [DataSet(F2-ArchiveMOCMA on f42 5-D), DataSet(F2-ArchiveMOCMA on f42 5-D)]
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
>     "__main__", fname, loader, pkg_name)
>   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code exec code in run_globals
>   File "/home/ulfi/.local/lib/python2.7/site-packages/bbob_pproc-0.0.0-py2.7.egg/bbob_pproc/__main__.py", line 298, in <module>  rungeneric.main(args)
>   File "/home/ulfi/.local/lib/python2.7/site-packages/bbob_pproc-0.0.0-py2.7.egg/bbob_pproc/rungeneric.py", line 287, in main
>     rungenericmany.main(genopts + ["-o", outputdir] + args)
>   File "/home/ulfi/.local/lib/python2.7/site-packages/bbob_pproc-0.0.0-py2.7.egg/bbob_pproc/rungenericmany.py", line 400, in main genericsettings.verbose)
>   File "/home/ulfi/.local/lib/python2.7/site-packages/bbob_pproc-0.0.0-py2.7.egg/bbob_pproc/compall/pptables.py", line 289, in main
>     raise Exception(txt)
> Exception: There is more than a single entry associated with folder ./exdata/Algorithm on 5-D f42.

I am trying to rewrite the code to locate all problems on the same function at the same observer

**Comment by Ulfgard** [comment]
Hi,

okay. i moved all problems involving the same function to the same observer. This works. Almost.

I did it like that:
```
std::vector<std::vector<coco_problem_t*> >problems(55);
for (std::size_t problem_index = 0; problem_index < coco_suite_get_number_of_problems(suite); problem_index++) {
	coco_problem_t* problem = coco_suite_get_problem(suite, problem_index);
	if (problem != NULL){ /* Skip the inaccessible problems if the suite was filtered */
		std::size_t function_idx, instance_idx, dimension_idx; //get function id
		coco_suite_decode_problem_index(suite, problem_index,&function_idx, &instance_idx,&dimension_idx);
		problems[function_idx].push_back(problem);
	}
}
```

the evaluation is as before only that i now start threads by looping over the function_idx and then process all problems in problems[function_idx] on the same thread.

This works and the post-processing does not complain. However, if i then run the suite over several dimensions i again run into trouble with the output of the *.info file. These read:

> algorithm = 'Algorithm', indicator = 'hyp', folder = '1-separable_1-separable'
> % My awesome algorithm
> function =  1, dim =  2, bbob-biobj_f01_d02_hyp.dat, 1:200000|1.4e-05, 2:200000|1.4e-05, 3:200000|1.4e-05, 4:200000|1.4e-05, 5:200000|1.4e-05, 6:200000|1.7e-01, 7:200000|1.7e-01, 8:200000|1.7e-01, 9:200000|1.7e-01, 10:200000|1.7e-01, 1:300000|1.2e-05, 2:300000|1.2e-05, 3:300000|1.1e-05, 4:300000|1.2e-05, 5:300000|1.2e-05, 6:300000|1.7e-01, 7:300000|1.7e-01, 8:300000|1.7e-01, 9:300000|1.7e-01, 10:300000|1.7e-01, 1:500000|4.7e-06, 2:500000|4.9e-06, 3:500000|4.7e-06, 4:500000|4.9e-06, 5:500000|5.2e-06, 6:500000|1.7e-01, 7:500000|1.7e-01, 8:500000|1.7e-01, 9:500000|1.7e-01, 10:500000|1.7e-01
> ...
they should read:

> algorithm = 'Algorithm', indicator = 'hyp', folder = '1-separable_1-separable'
> % My awesome algorithm
> function =  1, dim =  2, bbob-biobj_f01_d02_hyp.dat, 1:200000|1.4e-05, 2:200000|1.4e-05, 3:200000|1.4e-05, 4:200000|1.4e-05, 5:200000|1.4e-05, 6:200000|1.7e-01, 7:200000|1.7e-01, 8:200000|1.7e-01, 9:200000|1.7e-01, 10:200000|1.7e-01
> function =  1, dim =  3, bbob-biobj_f01_d02_hyp.dat, 1:300000|1.2e-05, 2:300000|1.2e-05, 3:300000|1.1e-05, 4:300000|1.2e-05, 5:300000|1.2e-05, 6:300000|1.7e-01, 7:300000|1.7e-01, 8:300000|1.7e-01, 9:300000|1.7e-01, 10:300000|1.7e-01
> function =  1, dim =  5, bbob-biobj_f01_d02_hyp.dat,  1:500000|4.7e-06, 2:500000|4.9e-06, 3:500000|4.7e-06, 4:500000|4.9e-06, 5:500000|5.2e-06, 6:500000|1.7e-01, 7:500000|1.7e-01, 8:500000|1.7e-01, 9:500000|1.7e-01, 10:500000|1.7e-01
> ...

All different trials are somehow treated as being the same dimensionality. Different functions are still treated as different, though. I fear that the observer somehow gets a wrong impression about the dimensionality. maybe it picks it up via the suite which is of course at 2 all the time.

**Comment by ttusar** [comment]
So does this mean that the parallelization is now working fine for you?

BTW, you might have received an e-mail from our continuous integration because of a failed test - ignore it :)

**Comment by Ulfgard** [comment]
Well, it works. But I don't know how good. I am a bit troubled that the order of function calls is so important. I would guess there are still bugs in the code, e.g. it must work when i, instead of deleting the problem, just remove the observer from it. It does not. Not even single threaded.

I am too busy to investigate what exactly is going wrong, but you might consider this as another open issue. For now, i managed to get it running.

**Comment by Ulfgard** [comment]
I stumbled over another issue, but i can not clearly identify whether it is the multi-threading that caused it.

In one of the runs, i get two result files:

bbob-biobj(bbob_f001_i02_d1
bbob-biobj_f48_d10_hyp.dat

where the first file contains exactly one run, while the other file contains the other runs. 
in the corresponding .info file also the file bbob-biobj\(bbob_f001_i02_d1 is referenced.  So i guess for some reason the wrong filename is generated in this single case

**Comment by ttusar** [comment]
I was assuming that parallelization of the experiments would be done by splitting the suite using `suite_options` rather than the code above, which is using a poorly tested approach. I would guess the problems that you describe stem from this. If you are able to share your code snippet, I can try to reproduce the problem and find a fix for it.

**Comment by nikohansen** [comment]
@ttusar, how would you filter the instances such that the predicted runtime over the sets are similar? `1-100`, `101-200` etc is not suitable. What should work is something like `[i for i in range(number_of_problems) if not i % i_subset]`, however how can we pass this as filter option?

**Comment by ttusar** [comment]
I do realize the drawbacks, or rather, limitations of using the existing suite options. We could (should) have new `batch` and `number_of_batches` suite options for this. It shouldn't be too difficult to do...
